zkevm-circuits/src/evm_circuit/execution/precompiles/modexp.rs

```diff
@@ -0,0 +1,1144 @@
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        big_int: &[u8; MODEXP_SIZE_LIMIT],
+    ) -> Result<(), Error> {
+        let byte14_lo = big_int[MODEXP_SIZE_LIMIT - 14] & 0xf;
+        let byte14_hi = big_int[MODEXP_SIZE_LIMIT - 14] & 0xf0;
+
+        self.byte14_split_lo
+            .assign(region, offset, Value::known(F::from(byte14_lo as u64)))?;
+        self.byte14_split_hi
+            .assign(region, offset, Value::known(F::from(byte14_hi as u64)))?;
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ModExpGadget<F> {
+    is_success: Cell<F>,
+    callee_address: Cell<F>,
+    caller_id: Cell<F>,
+    call_data_offset: Cell<F>,
+    call_data_length: Cell<F>,
+    return_data_offset: Cell<F>,
+    return_data_length: Cell<F>,
+
+    input: ModExpInputs<F>,
+    output: ModExpOutputs<F>,
+
+    input_bytes_acc: Cell<F>,
+    output_bytes_acc: Cell<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ModExpGadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::PrecompileBigModExp;
+
+    const NAME: &'static str = "MODEXP";
+
+    fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
+        // we 'copy' the acc_bytes cell inside call_op step, so it must be the first query cells
+        let input_bytes_acc = cb.query_cell_phase2();
+        let output_bytes_acc = cb.query_cell_phase2();
+
+        let [is_success, callee_address, caller_id, call_data_offset, call_data_length, return_data_offset, return_data_length] =
+            [
+                CallContextFieldTag::IsSuccess,
+                CallContextFieldTag::CalleeAddress,
+                CallContextFieldTag::CallerId,
+                CallContextFieldTag::CallDataOffset,
+                CallContextFieldTag::CallDataLength,
+                CallContextFieldTag::ReturnDataOffset,
+                CallContextFieldTag::ReturnDataLength,
+            ]
+            .map(|tag| cb.call_context(None, tag));
+
+        cb.precompile_info_lookup(
+            cb.execution_state().as_u64().expr(),
+            callee_address.expr(),
+            cb.execution_state().precompile_base_gas_cost().expr(),
+        );
+
+        let input = ModExpInputs::configure(cb, call_data_length.expr(), input_bytes_acc.expr());
+
+        let call_success = util::and::expr([
+            input.is_valid(),
+            //TODO: replace this constants when gas gadget is ready
+            1.expr(),
+        ]);
+
+        cb.require_equal(
+            "call success if valid input and enough gas",
+            is_success.expr(),
+            call_success.clone(),
+        );
+
+        let output = ModExpOutputs::configure(
+            cb,
+            output_bytes_acc.expr(),
+            //FIXME: there may be still some edge cases lead to nil output (even modulus_len is
+            // not 0)
+            call_success,
+            input.modulus_len(),
+        );
+
+        cb.condition(util::not::expr(output.is_output_nil()), |cb| {
```

Comment:
Cool. We should always check output word against output_bytes_rlc (if output is nil, output word would be assigned all zero), and constraint output_bytes_rlc is 0 when output is nil since there is no entry in modexp table for this case.

Both have been fixed in ModExpOutput gadget and unittest is added